pkg/binding/bindingpolicy.go

```diff
@@ -96,27 +103,26 @@ func (c *Controller) handleBindingPolicy(ctx context.Context, obj runtime.Object
 		return c.requeueWorkloadObjects(ctx, bindingPolicy.Name)
 	}
 
-	// handle deletion of bindingpolicy
-	return c.handleBindingPolicyDeletion(ctx, bindingPolicy)
+	// we delete finalizer if the policy is being deleted (not yet deleted).
+	if err = c.handleBindingPolicyFinalizer(ctx, bindingPolicy); err != nil {
+		return fmt.Errorf("failed to handle finalizer for bindingPolicy %s: %w", bindingPolicy.Name, err)
+	}
+
+	return c.updateResolutionForDeletedBindingPolicy(ctx, objIdentifier.ObjectName.Name)
 }
 
-func (c *Controller) handleBindingPolicyDeletion(ctx context.Context, bindingPolicy *v1alpha1.BindingPolicy) error {
-	if c.bindingPolicyResolver.ResolutionRequiresSingletonReportedState(bindingPolicy.GetName()) {
+func (c *Controller) updateResolutionForDeletedBindingPolicy(ctx context.Context, bindingPolicyName string) error {
+	if c.bindingPolicyResolver.ResolutionRequiresSingletonReportedState(bindingPolicyName) {
 		// if the bindingpolicy required a singleton status, all selected objects should
 		// be requeued in order to remove the label
-		if err := c.requeueSelectedWorkloadObjects(ctx, bindingPolicy.GetName()); err != nil {
+		if err := c.requeueSelectedWorkloadObjects(ctx, bindingPolicyName); err != nil {
```

Comment:
FYI: There can be multiple updates to a BindingPolicy while it is in the process of being removed.

Reply:
After the first one goes through, the resolution would be deleted, and this block would not be entered.

Reply:
Ah, right.